### PR TITLE
Fix headers to be handled case insensitive and ignore 100-continue for HTTP/1.0

### LINKED ISSUE #36: Remove X-Powered-By: React/alpha or make optional
For security reasons, among others, we should remove the `X-Powered-By` header.

### LINKED ISSUE #51: Header names are not case-sensitive
https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2

`Each header field consists of a name followed by a colon (":") and the field value. Field names are case-insensitive`

Currently the [code](https://github.com/reactphp/http/blob/master/src/RequestParser.php#L67) [heavily](https://github.com/reactphp/http/blob/master/src/RequestParser.php#L127) relies on case-sensitive names.

Also Content-Type can contain a charset or additional information after `;` character. Example: `Content-Type: application/x-www-form-urlencoded; charset=utf-8` Such header break currently content parsing because of [this](https://github.com/reactphp/http/blob/master/src/RequestParser.php#L141).

We tried to fix it in PPM: https://github.com/php-pm/php-pm/blob/master/React/RequestParser.php#L18, but I believe we should fix it directly in reactphp/http.

### PR DESCRIPTION
**Pull request by clue**
This simple PR makes sure we always match header names case insensitive.

Also adds some documentation for the existing methods.

Builds on top of #103 
Fixes / closes #51
Fixes / closes #36
Supersedes / closes #54
Supersedes #53
Supersedes #47

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -79,10 +84,110 @@ Returns a comma-separated list of all values for this header name or an empty st
...
+client from transfering the whole body at all.
+
+This method is mostly useful in combination with the
+[`expectsContinue()`](#expectscontinue) method like this:
+
+```php
+$http->on('request', function (Request $request, Response $response) {
+    if ($request->expectsContinue()) {
+        $response->writeContinue();
+    }
```
Comment (kelunik) [inline README.md]: How can I access the actual request body then? I think this should be handled transparently.

File: src/Response.php [hunk 2]
```diff
@@ -46,21 +85,87 @@ public function writeContinue()
...
-            array('X-Powered-By' => 'React/alpha'),
-            $headers
-        );
+        // assign default "X-Powered-By" header as first for history reasons
+        if (!isset($lower['x-powered-by'])) {
+            $headers = array_merge(
+                array('X-Powered-By' => 'React/alpha'),
+                $headers
+            );
+        }
```
Comment (kelunik) [inline src/Response.php]: This should probably update `$lower`, too?

File: README.md (hunk 1 above)
Comment (clue) [inline README.md]: This only adds documentation for the existing behavior, accessing the request body is not part of this :+1: (see also #104 and #105)

That being said and given the current architecture, I think it's very reasonable to handle this explicitly. The documentation explicitly states why the consumer of this lib is in control of this behavior. (see also #47)

Note that I'm not saying this has to stay that way, but I'd rather discuss this in a separate feature ticket instead (BC break ahead). Do you feel like filing one? :+1:

File: src/Response.php (hunk 2 above)
Comment (clue) [inline src/Response.php]: The `$lower` is only used within this method, so it's not needed anymore after this assignment.

File: README.md (hunk 1 above)
Comment (kelunik) [inline README.md]: If I write code like documented now, how does reading the request body work then? Do I have to find it out manually and discard the first expect 100 request?

Comment (clue) [inline README.md]: See the existing and the updated documentation: The `Request` implements `ReadableStreamInterface`, so you can read the request body by awaiting the `data` events and wait for a final `end` event.

Note that this is somewhat limited until #104 is resolved, which is something I'm currently looking into with @legionth :+1: 

Note that this is actually only a *single* request, so nothing is being "discarded" here.

File: src/Request.php [hunk 3]
```diff
@@ -111,9 +111,19 @@ public function hasHeader($name)
...
+     * request body.
+     * See [`writeContinue()`] for more details.
+     *
+     * @return bool
+     * @see Response::writeContinue()
+     */
     public function expectsContinue()
     {
-        return isset($this->headers['Expect']) && '100-continue' === $this->headers['Expect'];
+        return '100-continue' === $this->getHeaderLine('Expect');
```
Comment (kelunik) [inline src/Request.php]: This needs a `strtolower` call for `$this->getHeaderLine("expect")`, see https://tools.ietf.org/html/rfc7231#section-5.1.1

> The Expect field-value is case-insensitive.

Comment (kelunik) [inline src/Request.php]: This should also check the protocol version, as it must return `false` for HTTP/1.0.

> A server that receives a 100-continue expectation in an HTTP/1.0 request MUST ignore that expectation.

Comment (clue) [inline src/Request.php]: Good catch, thanks! :+1:

File: README.md (hunk 1 above)
Comment (kelunik) [inline README.md]: @clue Remembered it wrong, thought expect-100 would send the request headers again then.

**Comment by clue** [comment]
Ping, afaict everything's been resolved :shipit:

**Pull request merged**